File: donghyo/baekjoon/2178.py
from collections import deque

def bfs(N, M, matrix, i, j, visited):
    validx, validy = [-1, 1, 0, 0], [0, 0, -1, 1]
    
    queue = deque()
    queue.append([i,j])

    visited[i][j] = 1

    while queue:
      x, y  = queue.popleft()

      if x == N - 1 and y == M - 1:
         return visited[x][y]  
          
      print(visited)

      for i in range(4):
            check_x, check_y = x + validx[i], y + validy[i]
            if 0 <= check_x < N and 0 <= check_y < M:
                if matrix[check_x][check_y] == 1:
                    if visited[check_x][check_y] == 0:
                        visited[check_x][check_y] = visited[x][y] + 1
                        queue.append([check_x,check_y])          

    
def main():
    N, M = list(map(int, input().split()))
    matrix =[list(map(int, input())) for _ in range(N)]
    visited = [[0] * M for _ in range(N)]
    answer = 0
    answer = bfs(N, M, matrix, 0, 0, visited)
    print(answer)

main()

File: donghyo/baekjoon/2178.js
// 미로탐색

let solution = (matrix, N, M) => {
    let [valid_x, valid_y] = [[-1, 1, 0, 0], [0, 0, -1, 1]];
    let visited = Array.from(Array(N), () => Array());
    let queue = [];

    for (let i = 0; i < N; i++) {
        for (let j = 0; j < M; j++) {
            visited[i][j] = 0;
        }
    }
    visited[0][0] = 1;
    queue.push([0, 0]);

    while (queue.length) {
        [x, y] = queue.shift();

        if (x == N - 1 && y == M - 1) {
            return visited[x][y];
        }

        for (let i = 0; i < 4; i++) {
            let check_x = x + valid_x[i];
            let check_y = y + valid_y[i];

            if (0 <= check_x && check_x < N && 0 <= check_y && check_y < M) {
                if (visited[check_x][check_y] == 0 && matrix[check_x][check_y] == 1) {
                    visited[check_x][check_y] = visited[x][y] + 1
                    queue.push([check_x, check_y]);
                }
            }
        }
    }
}

let func = () => {
    let N = 4, M = 6;
    matrix = [[101111], [101010], [101011], [111011]];

    splitMatrix = [];
    for (let i = 0; i < matrix.length; i++) {
        for (let j = 0; j < matrix[i].length; j++) {
            splitMatrix.push(String(matrix[i]).split(''))
        }
    }

    let result = solution(splitMatrix, N, M);
    console.log(result);

}
func();

